group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -728,6 +794,81 @@ public void replay(
             }
             consumerGroup.removeMember(memberId);
         }
+
+        updateGroupsByTopics(groupId, oldSubscribedTopicNames, consumerGroup.subscribedTopicNames());
```

Comment:
`groupsByTopics` is a timeline data structure and we update it based on ConsumerGroupMemberMetadataKey/Value record. If the record can't be written, the state is reverted to the last written offset and this reverts all those pending changes.

> i think i'm confused because in api handling (i.e. consumer group heartbeat) once we modify the timeline data structures we generate records to commit the offset in the timeline but here we do it in reverse

In the api handling, we never update the timeline data structures. We only generate records and they are updated when the records are applied.